src/data/simple_csr_source.cc

```diff
@@ -160,100 +174,35 @@ const SparsePage& SimpleCSRSource::Value() const {
-      CHECK_EQ(j_shape.size(), 1) << ColumnarErrors::Dimension(1);
-      CHECK_EQ(get<Integer>(j_shape.front()) % 8, 0) <<
-          "Length of validity map must be a multiple of 8 bytes.";
-      int64_t size = get<Integer>(j_shape.at(0)) *
-                     sizeof(unsigned char) / sizeof(RBitField8::value_type);
-      s_mask = {p_mask, size};
-      auto typestr = get<String const>(j_mask.at("typestr"));
-      CHECK_EQ(typestr.size(),    3) << ColumnarErrors::TypestrFormat();
-      CHECK_NE(typestr.front(), '>') << ColumnarErrors::BigEndian();
-      CHECK_EQ(typestr.at(1),   'i') << "mask" << ColumnarErrors::ofType("unsigned char");
-      CHECK_EQ(typestr.at(2),   '1') << "mask" << ColumnarErrors::toUInt();
-
-      CHECK(j_mask.find("null_count") != j_mask.cend()) <<
-          "Column with null mask must include null_count as "
-          "part of mask object for XGBoost.";
-      null_count = get<Integer const>(j_mask.at("null_count"));
-    }
+  CHECK_GT(n_columns, 0) << "Number of columns must not be greater than 0.";
 
-    // Find data field
-    if (column.find("data") == column.cend()) {
-      LOG(FATAL) << "Empty dataset passed in.";
-    }
-
-    auto typestr = get<String const>(column.at("typestr"));
-    CHECK_EQ(typestr.size(),    3) << ColumnarErrors::TypestrFormat();
-    CHECK_NE(typestr.front(), '>') << ColumnarErrors::BigEndian();
-    CHECK_EQ(typestr.at(1),   'f') << "data" << ColumnarErrors::ofType("floating point");
-    CHECK_EQ(typestr.at(2),   '4') << ColumnarErrors::toFloat();
-
-    auto j_shape = get<Array const>(column.at("shape"));
-    CHECK_EQ(j_shape.size(), 1) << ColumnarErrors::Dimension(1);
-
-    if (column.find("strides") != column.cend()) {
-      auto strides = get<Array const>(column.at("strides"));
-      CHECK_EQ(strides.size(), 1)              << ColumnarErrors::Dimension(1);
-      CHECK_EQ(get<Integer>(strides.at(0)), 4) << ColumnarErrors::Contigious();
-    }
-
-    auto length = get<Integer const>(j_shape.at(0));
-
-    float* p_data = GetPtrFromArrayData<float*>(column);
-    common::Span<float> s_data {p_data, length};
-
-    foreign_cols[i].data  = s_data;
-    foreign_cols[i].valid = RBitField8(s_mask);
-    foreign_cols[i].size  = s_data.size();
-    foreign_cols[i].null_count = null_count;
-  }
+  auto const& typestr = get<String const>(columns[0]["typestr"]);
+  CHECK_EQ(typestr.size(),    3)  << ColumnarErrors::TypestrFormat();
+  CHECK_NE(typestr.front(), '>')  << ColumnarErrors::BigEndian();
 
-  info.num_col_ = n_columns;
-  info.num_row_ = foreign_cols[0].size;
-  for (size_t i = 0; i < n_columns; ++i) {
-    CHECK_EQ(foreign_cols[0].size, foreign_cols[i].size);
-    info.num_nonzero_ += foreign_cols[i].data.size() - foreign_cols[i].null_count;
+  for (auto const& col : columns) {
+    auto const& typestr_i = get<String const>(col["typestr"]);
+    CHECK_EQ(typestr, typestr_i) << "Columns should be of same type.  Preferably float32.";
```

Comment:
How hard is it to relax this constraint? I can see common scenarios with mixed data types.